File: Advanced/Section/CustomDrillThrough/C#/CustomDrillThrough/ViewerForm.cs
using System;
using System.Windows.Forms;
using System.Xml;

namespace GrapeCity.ActiveReports.Samples.CustomDrillThrough
{
	/// <summary>
	///Summary description for frmViewMain. 
	/// 
	///Demonstrates using hyperlinks and the viewer hyperlink event to simulate  
	///drill-through from one report to another.
	/// </summary>
	public partial class ViewerMain : System.Windows.Forms.Form
	{
		/// <summary>
		/// Required designer variable.
		/// </summary>
		private readonly bool _loadMainReport;

		public ViewerMain()
		{
			//
			// Required for Windows Form Designer support
			//
			InitializeComponent();
			_loadMainReport = true;
		}

		public ViewerMain(bool loadMainReport)
		{
			InitializeComponent();
			_loadMainReport = loadMainReport;
		}

		/// <summary>
		/// The main entry point for the application.
		/// 
		/// </summary>
		[STAThread]
		static void Main() 
		{
			Application.Run(new ViewerMain());
		}

		/// <summary>
		/// Handles the hyperlink sent from the viewer object.
		/// </summary>
		private void arvMain_HyperLink(object sender, GrapeCity.ActiveReports.Viewer.Win.HyperLinkEventArgs e)
		{
			e.Handled = true;
			string hyperlink = e.HyperLink.Split(':')[1];
			string report = e.HyperLink.Split(':')[0];

			if (report == "DrillThrough1")
			{
				// Click on customer ID to open the drill through for that customer.
				var rpt2 = new SectionReport();
				rpt2.LoadLayout(XmlReader.Create(Properties.Resources.DrillThrough1));
				ViewerMain frm2 = new ViewerMain(false);
				rpt2.Parameters["customerID"].Value = hyperlink;
				frm2.arvMain.LoadDocument(rpt2);
				frm2.ShowDialog(this);
			}
			else if (report == "DrillThrough2")
			{
				// Click order number to open the order details
				var rpt3 = new SectionReport();
				rpt3.LoadLayout(XmlReader.Create(Properties.Resources.DrillThrough2));
				ViewerMain frm3 = new ViewerMain(false);
				rpt3.Parameters["orderID"].Value = hyperlink;
				frm3.arvMain.LoadDocument(rpt3);
				frm3.ShowDialog(this);
			}
		}

		/// <summary>
		/// ViewerForm_Load - Handles initial form setup.
		/// </summary>
		private void ViewerForm_Load(object sender, EventArgs e)
		{
			if(_loadMainReport)
			{
				var rpt = new SectionReport();
				rpt.LoadLayout(XmlReader.Create(Properties.Resources.DrillThroughMain));
				arvMain.LoadDocument(rpt);
			}
		}
	}
}

File: Advanced/Section/CustomDrillThrough/VB.NET/CustomDrillThrough/ViewerForm.vb
Imports System
Imports System.Windows.Forms
Imports System.Xml

Public Class ViewerForm
	Inherits System.Windows.Forms.Form

	Private _loadMainReport As Boolean

	Public Sub New()
		MyBase.New()

		InitializeComponent()
		_loadMainReport = True
	End Sub

	Public Sub New(ByVal loadMainReport As Boolean)
		MyBase.New()
		InitializeComponent()
		_loadMainReport = loadMainReport
	End Sub

	Protected Overloads Overrides Sub Dispose(ByVal disposing As Boolean)
		If disposing Then
			If Not (components Is Nothing) Then
				components.Dispose()
			End If
		End If
		MyBase.Dispose(disposing)
	End Sub

	<STAThread()>
	Shared Sub Main()
		Application.Run(New ViewerForm())
	End Sub 'Main

	' arvMain_HyperLink: Handles the hyperlink sent from the viewer object.
	Private Sub arvMain_HyperLink(ByVal sender As Object, ByVal e As Viewer.Win.HyperLinkEventArgs) Handles arvMain.HyperLink
		e.Handled = True
		Dim hyperlink As String = e.HyperLink.Split(New String() {":"}, StringSplitOptions.None)(1)
		Dim report As String = e.HyperLink.Split(New String() {":"}, StringSplitOptions.None)(0)
		If report = "DrillThrough1" Then
			' Click on customer ID to open the drill through for that customer.
			Dim rpt2 As New SectionReport
			rpt2.LoadLayout(XmlReader.Create(My.Resources.DrillThrough1))
			Dim frm2 As New ViewerForm(False)
			rpt2.Parameters("customerID").Value = hyperlink
			frm2.arvMain.LoadDocument(rpt2)
			frm2.ShowDialog(Me)
		ElseIf report = "DrillThrough2" Then
			' Click order number to open the order details
			Dim rpt3 As New SectionReport
			rpt3.LoadLayout(XmlReader.Create(My.Resources.DrillThrough2))
			Dim frm3 As New ViewerForm(False)
			rpt3.Parameters("orderID").Value = hyperlink
			frm3.arvMain.LoadDocument(rpt3)
			frm3.ShowDialog(Me)
		End If
	End Sub 'arvMain_HyperLink

	' ViewerForm_Load - Handles initial form setup.
	Private Sub Form1_Load(ByVal sender As Object, ByVal e As EventArgs) Handles MyBase.Load
		If _loadMainReport Then
			Dim rpt As New SectionReport
			rpt.LoadLayout(XmlReader.Create(My.Resources.DrillThroughMain))
			arvMain.LoadDocument(rpt)
		End If
	End Sub 'Form1_Load

End Class 'ViewerForm
